File: Semester 4/Class12_Computer_Contest_Level_3_Notes_and_Homework_Nov_26/lvl3_hw_sol/largest_min_distance.cpp
#include <bits/stdc++.h>
using namespace std;
const int MM = 1e5+5;
int N, M, p[MM], ans;
bool check(int lmt){
    int last=p[1], cnt = 1;
    for(int i=2; i<=N; i++){
        if(p[i] - last >= lmt){
            last = p[i]; cnt++;
        }
    }
    return cnt >= M;
}
int main(){
    cin >> N >> M;
    for(int i=1; i<=N; i++) cin >> p[i];
    sort(p+1, p+N+1);
    int lo = 1, hi = p[N] - p[1];
    while(lo <= hi){
        int mid = (lo + hi)/2;
        if( check(mid) ) { ans = mid; lo = mid + 1;}
        else hi = mid - 1;
    }
    cout << ans << endl;
}

File: Semester 4/Class12_Computer_Contest_Level_3_Notes_and_Homework_Nov_26/lvl3_hw_sol/largest_min_distance.java
import java.io.*;
import java.time.Period;
import java.util.*;

public class homework {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int n, m, a[];
    public static void main(String[] args) throws IOException {
    	n = readInt(); m = readInt(); a = new int[n];
    	for(int i=0; i<n; i++) a[i] = readInt();
    	Arrays.sort(a);
    	int lo = 1, hi = a[n-1], ans = 0;
    	while(lo <= hi) {
    		int mid = (lo + hi)/2;
    		if(check(mid)) { ans = mid; lo = mid + 1; }
    		else hi = mid - 1;
    	}
    	System.out.println(ans);
    }
    static boolean check(int x) {
    	int last = a[0], cnt = 1;
    	for(int i=1; i<n; i++) {
    		if(a[i] - last >= x) {
    			last = a[i]; cnt++;
    		}
    	}
    	return cnt >= m;
    }
    static String next() throws IOException {
        while (st == null || !st.hasMoreTokens())
            st = new StringTokenizer(br.readLine().trim());
        return st.nextToken();
    }

    static long readLong() throws IOException {
        return Long.parseLong(next());
    }

    static int readInt() throws IOException {
        return Integer.parseInt(next());
    }

    static double readDouble() throws IOException {
        return Double.parseDouble(next());
    }

    static char readCharacter() throws IOException {
        return next().charAt(0);
    }

    static String readLine() throws IOException {
        return br.readLine().trim();
    }
}